rainfall_runoff_IHACRES.R

```diff
@@ -0,0 +1,241 @@
+  return(r_squared)
+}
+
+
+################################################################################
+# Run optimization
+
+# Some optimization references
+#https://www.openanalytics.eu/blog/2022/05/17/new-nonlinear-ls-solvers-in-r/
+
+#' method 1: 
+#' stats::optim - General-purpose optimization based on Nelder–Mead, 
+#' quasi-Newton and conjugate-gradient algorithms. It includes an option for 
+#' box-constrained optimization and simulated annealing.
+
+# tau_q, tau_s, vs, delay, loss
+optim_par <- optim(fn = fit_fun,
+           method = 'L-BFGS-B',
+           par = c(3, 50, 0.5, 24, 0.5), 
+           lower = c(0.1, 50, 0.01, 1, 0.01),
+           upper = c(20, 120, .8, 50, 1),
+           control = c("trace" = 6, "factr" = .001),
+           obj = "MAE")# function
+# Plot results
+print(optim_par$par)
+ggarrange(fit_fun(params = optim_par$par, plot = TRUE, return_dat = FALSE),
+          fit_fun(params = optim_par$par, plot = TRUE, return_dat = FALSE, 
+                  dat = "test_data"), nrow = 2)
+
+
+#' Method 2:
+#' stats::nlminb - Unconstrained and box-constrained optimization using PORT 
+#' routines.
+nlminb_par <- nlminb(start = c(3, 50, 0.5, 24, 0.5), 
+            fit_fun,
+            lower = c(0.1, 50, 0.01, 1, 0.01),
+            upper = c(20, 120, .8, 50, 1),
+            control = list("trace" = 6, "iter.max" = 200))
+
+print(nlminb_par$par)
+ggarrange(fit_fun(params = nlminb_par$par, plot = TRUE, return_dat = FALSE),
+          fit_fun(params = nlminb_par$par, plot = TRUE, return_dat = FALSE, 
+                  dat = "test_data"), nrow = 2)
+
+#' Method 3: 
+#' Manual
+manual_par <- c(20,90,.6,15,.5)
+Q_mod <- fit_fun(params = manual_par, plot = TRUE, return_dat = TRUE)
+
+################################################################################
+# Exploring results
+
+
+# Example plotting outside of fit_fun
+
+autoplot(cbind(cal_data$Q,Q_mod$Q_mod),
+         facet=FALSE,
+         ylab = "Discharge (in/d)",xlab="year") +
+  scale_color_manual("",
+                     values = c("dodgerblue","black"),
+                     labels = c("observed", "modeled"))
+
+
+# Looking at recharge with calibrated model
+# set which pars to use (optim_par, nlminb_par, manual_par)
+params <- optim_par$par
+mod_out <- ihacres_routing_gw(U = U_in, 
+                   tau_q = params[1], 
+                   tau_s = params[2], 
+                   vs = params[3], 
+                   delay = as.integer(params[4]), 
+                   loss = params[5],
+                   return_components = TRUE)
+
+autoplot(cbind(cal_data$Q, mod_out$Q_mod, mod_out$Xg), facet = FALSE) +
+  scale_color_manual("",values = c("black","dodgerblue","hotpink"))
+
+
+# Q-Q plot of residuals
+# this doesn't look great... should be a linear line if the residuals are evenly
+# distributed.. need to work through different objective functions to consider
+# how each one biases results. 
+qqnorm(mod_out$Q_mod - cal_data$Q)
```

Comment:
Wonder if transforming Q (log10, square root, etc.) ahead of model dev could help with this?